File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/controller/dto/Virksomhet.java
package no.nav.melosys.eessi.controller.dto;

import lombok.Data;
import no.nav.melosys.eessi.models.sed.nav.Arbeidsgiver;
import no.nav.melosys.eessi.models.sed.nav.Identifikator;
import no.nav.melosys.eessi.service.sed.helpers.StreamUtils;

@Data
public class Virksomhet {

    private String navn;
    private Adresse adresse;
    private String orgnr;
    private String type; //Trenger kanskje ikke denne?

    public static Virksomhet av(Arbeidsgiver arbeidsgiver) {
        Virksomhet virksomhet = new Virksomhet();

        virksomhet.navn = arbeidsgiver.getNavn();
        virksomhet.adresse = Adresse.av(arbeidsgiver.getAdresse());
        virksomhet.orgnr = StreamUtils.nullableStream(arbeidsgiver.getIdentifikator())
                .findFirst().map(Identifikator::getId).orElse(null);

        return virksomhet;
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/controller/dto/Virksomhet.kt
package no.nav.melosys.eessi.controller.dto

import no.nav.melosys.eessi.models.sed.nav.Arbeidsgiver
import no.nav.melosys.eessi.models.sed.nav.Identifikator
import java.util.stream.Stream

data class Virksomhet(
    var navn: String? = null,
    var adresse: Adresse,//? = null, // Kaster NullPointereException i java kode
    var orgnr: String? = null,
    var type: String? = null //Trenger kanskje ikke denne?
) {
    companion object {
        fun av(arbeidsgiver: Arbeidsgiver): Virksomhet {
            return Virksomhet(
                navn = arbeidsgiver.navn,
                adresse = arbeidsgiver.adresse.let { Adresse.av(it) },
                orgnr = nullableStream(arbeidsgiver.identifikator)
                    .findFirst()
                    .map(Identifikator::id)
                    .orElse(null)
            )
        }

        fun <T> nullableStream(collection: Collection<T>?): Stream<T> = collection?.stream() ?: Stream.empty()
    }
}
